tools/clang/unittests/HLSL/ExecutionTest.cpp

```diff
@@ -1272,6 +1283,263 @@ void ExecutionTest::RunRWByteBufferComputeTest(ID3D12Device *pDevice, LPCSTR pSh
+      const int val = g_bab.Load(addr);
+      int res = val;
+      if (val < 0) { // Never true.
+        int arr[200];
+        for (int i = 0; i < 200; ++i) {
+            arr[i] = arr[val - i];
+        }
+        res += arr[val];
+      }
+      g_bab.Store(addr, (uint)res);
+    }
+
+    [numthreads(8,8,1)]
+    void main(uint GI : SV_GroupIndex) {
+      fn(GI);
+    }
+
+    [shader("raygeneration")]
+    void RayGen() {
+      const uint d = DispatchRaysIndex().x;
+      const uint g = g > 64 ? 63 : g;
+      fn(g);
+    }
+  )";
+  static const int NumThreadsX = 8;
+  static const int NumThreadsY = 8;
+  static const int NumThreadsZ = 1;
+  static const int ThreadsPerGroup = NumThreadsX * NumThreadsY * NumThreadsZ;
+  static const int DispatchGroupCount = 1;
+
+  // TODO: There's probably a lot of things in the rest of this test that could be stripped away.
+
+  CComPtr<ID3D12Device5> pDevice;
+  if (!CreateDevice(reinterpret_cast<ID3D12Device**>(&pDevice), D3D_SHADER_MODEL_6_5, true, true)) // TODO: We need 6.6!
+    return;
+
+  std::vector<uint32_t> values;
+  SetupComputeValuePattern(values, ThreadsPerGroup * DispatchGroupCount);
+
+  // Run a number of tests for different configurations that will cause
+  // lifetime intrinsics to be placed directly, be replaced by a zeroinitializer
+  // store, or be replaced by an undef store.
+  LPCWSTR pOptions15[] = {L"/validator-version 1.5"};
+  LPCWSTR pOptions16[] = {L"/validator-version 1.6", L"/Vd"};
+
+  VERIFY_ARE_EQUAL(values[0], (uint32_t)0);
+
+  // Test regular shader with zeroinitializer store.
+  RunLifetimeIntrinsicTest(pDevice, pShader, D3D_SHADER_MODEL_6_0, false, pOptions15, _countof(pOptions15), values);
+  VERIFY_ARE_EQUAL(values[0], (uint32_t)0);
+
+  // Test library with zeroinitializer store.
+  RunLifetimeIntrinsicTest(pDevice, pShader, D3D_SHADER_MODEL_6_3, true, pOptions15, _countof(pOptions15), values);
+  VERIFY_ARE_EQUAL(values[0], (uint32_t)0);
+
+  // Testing SM 6.6 and validator version 1.6 requires experimental shaders.
+  // being turned on.
+  if (!m_ExperimentalModeEnabled)
+      return;
+
+  // Test regular shader with undef store.
+  RunLifetimeIntrinsicTest(pDevice, pShader, D3D_SHADER_MODEL_6_0, false, pOptions16, _countof(pOptions16), values);
+  VERIFY_ARE_EQUAL(values[0], (uint32_t)0);
+
+  // Test library with undef store.
+  RunLifetimeIntrinsicTest(pDevice, pShader, D3D_SHADER_MODEL_6_3, true, pOptions16, _countof(pOptions16), values);
+  VERIFY_ARE_EQUAL(values[0], (uint32_t)0);
+
+  // Test regular shader with lifetime intrinsics.
+  RunLifetimeIntrinsicTest(pDevice, pShader, D3D_SHADER_MODEL_6_5, false, pOptions16, _countof(pOptions16), values); // TODO: Test 6.6 here!
+  VERIFY_ARE_EQUAL(values[0], (uint32_t)0);
+
+  // Test library with lifetime intrinsics.
+  RunLifetimeIntrinsicTest(pDevice, pShader, D3D_SHADER_MODEL_6_5, true, pOptions16, _countof(pOptions16), values); // TODO: Test 6.6 here!
+  VERIFY_ARE_EQUAL(values[0], (uint32_t)0);
```

Comment:
This ExecutionTest looks really good and I appreciate how easy it is to read.

The only suggestion I'd make is to not rely on zero as an indicator of success. So many failures can result in memory being zero which might lead to a false positive, I would recommend avoiding it. 

Even something only slightly more complex like comparing `values[1]` to `1` would make me feel better.

Reply:
Thanks Greg! I've updated the test with your suggestion.
In general, the test result doesn't really matter. The test is only about the runtime/driver accepting the generated code (store zeroinitializer, store undef, or lifetime intrinsics depending on the SM/validator version). I saw the pattern of testing against value 0 in the test from which I copied, so just stuck with that :).